static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,348 @@
+
+If an HTML element causes a no-CORS fetch, you can use the badly-named `crossorigin` attribute to switch it to a CORS request.
+
+```html
+<img crossorigin src="…" />
+<script crossorigin src="…"></script>
+<link crossorigin rel="stylesheet" href="…" />
+<link crossorigin rel="preload" as="font" href="…" />
+```
+
+When you switch these over to CORS, you get more visibility into the cross-origin resource:
+
+- You can paint the `<img>` to a `<canvas>` and read back the pixels.
+- You get more detailed stack traces for script.
+- You can explore the parsed stylesheet via [`link.sheet`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet).
+
+With `<link rel="preload">`, you need to ensure it uses CORS if the eventual request will also use CORS, otherwise it won't match in the preload cache.
+
+# CORS requests
+
+By default, a cross-origin CORS request is made without credentials. So, no cookies, no client certs, and no automatic `Authorization` header. However, same-origin requests include credentials.
+
+By the time CORS was developed, the `Referer` header was frequently spoofed or removed by browser extensions, so a new header, `Origin` was created, which provides the origin of the page that made the request.
+
+`Origin` is generally useful, so it's been added to lots of other types of request, such as websocket and `POST` requests. We tried adding it to regular `GET` requests too, but it broke a bunch of sites that assumed the presence of the `Origin` header means it's a CORS request 😬. Maybe one day.
+
+# CORS responses
+
+To pass the CORS check and give the other origin access to the response, the response must include this header:
+
+```
+Access-Control-Allow-Origin: *
+```
+
+The `*` can be replaced with the value of the request's `Origin` header, but `*` works for any requesting origin provided the request is sent without credentials (more on that in a bit).
+
+This give the other origin access to the response body, and also a subset of the headers:
+
+- `Cache-Control`
+- `Content-Language`
+- `Content-Type`
+- `Expires`
+- `Last-Modified`
+- `Pragma`
+
+The response can include another header, `Access-Control-Expose-Headers`, to reveal additional headers, as long as they're not part of the [💀 FORBIDDEN LIST 💀](https://fetch.spec.whatwg.org/#forbidden-header-name), which are always hidden for security reasons.
+
+```
+Access-Control-Expose-Headers: Custom-Header-1, Custom-Header-2
+```
+
+# Is it safe to expose resources via CORS?
+
+`Access-Control-Allow-Origin: *` only grants response visibility if the request is made without credentials, so it's totally safe to use on all resources _unless_ that resource contains private data that's 'secured' using something other than browser credentials.
```

Comment:
There's a problem here in Chrome apparently if you don't use `Vary`. See https://github.com/whatwg/fetch/issues/307

Reply:
Ah yes, I should probably have a section on `Vary`, since that catches people out.